### ISSUE TITLE
ZMQ and latency issue?

### ISSUE THREAD

**Issue by mimianddaniel** [saltstack/salt#27339]
we have about 13000 minions, and while performing regression testing from a master of master, i noticed high lagging in the event arriving to syndics residing other DC from master of master. As a result, the minions registered to that syndics also receives the event late.  I dont see any latency issue to the syndics in the same DC as master of master, is this caused by ZMQ or salt itself? 

``` yaml
           Salt: 2015.5.0
         Python: 2.7.10 (default, Jun  3 2015, 16:19:05)
         Jinja2: 2.8
       M2Crypto: 0.22
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.7.0
           RAET: Not Installed
            ZMQ: 4.1.2
           Mako: Not Installed
        Tornado: Not Installed
        timelib: Not Installed
       dateutil: 2.2
```

**Comment by jfindlay** [comment]
@mimianddaniel, I have spoken with other people dealing with similar issues, such as @plastikos.  It seems like a zmq queuing issue happening at the syndic, but that is about all I know.  @cachedout may also have some more input.

**Comment by plastikos** [comment]
We have several locations across the globe that we connected together with local masters and then syndics for a single master-of-masters.  That makes for ~30,000 minions total with ~8,000 at the largest site.  When publishing a test.ping from the MOM it completes at each of the locations (the local masters have all of the minion responses) in 60 to 120 seconds.  It then takes up to 10 minutes for all of the responses to be returned to the MOM.

I believe this roll-up time is due to ZMQ overhead delivering responses with inter-site latencies amplifying delivery time.  I have not focused on this problem yet due to other ZMQ priorities.  Here are some expected lines of investigation:

   * Custom build ZMQ with larger settings for ``out_batch_size``/``in_batch_size`` (8192 seems quite small for being able to batch messages)
   * Investigate ``max_io_events`` ZMQ compile-time setting
   * Add multiple syndic workers for parallel deliver of minion responses

Right now I am working with a customized ZMQ that provides diagnostic information on the contexts, sockets and pipes providing HWM, LWM, fill depth, memory usage and a few other details.  I expect to use it to characterize and measure syndic+MOM changes.

**Comment by austinpapp** [comment]
@plastikos i know this question is rather old at this point but im highly interested in what you've done here. do you have any other updates on this space?

**Comment by plastikos** [comment]
@austinpapp - at this time we stopped using ``syndic`` and operate the different sites independently.

Back when I was actively investigating I built an instrumented version of ZMQ that provided statistics on the queue depth and how messages drained through the systems.  I was uncomfortable about the algorithm ZMQ uses to buffer and flush messages (high/low watermark managing) - I'm not convinced that the design actually implements a high/low watermark and that at large scale the flow-control algorithm gets exposed as being incorrect by the observed symptoms.
